src/circuit_builder.rs

```diff
@@ -26,21 +26,71 @@ pub enum PublicInput {
 }
 
 impl PublicInput {
-    /// Returns the value of a PublicInput struct
-    pub fn value(&self) -> Vec<BlsScalar> {
+    /// Returns the serialized-size of the `PublicInput` structure.
+    pub const fn serialized_size() -> usize {
+        33usize
+    }
+
+    /// Returns the byte-representation of a [`PublicInput`].
+    /// Note that the underlying variants of this enum have different
+    /// sizes on it's byte-representation. Therefore, we need to return
+    /// the biggest one to set it as the default one.
+    pub fn to_bytes(&self) -> [u8; Self::serialized_size()] {
+        let mut bytes = [0u8; Self::serialized_size()];
         match self {
-            PublicInput::BlsScalar(scalar, _) => vec![*scalar],
-            PublicInput::JubJubScalar(scalar, _) => vec![BlsScalar::from(*scalar)],
-            PublicInput::AffinePoint(point, _, _) => vec![point.get_x(), point.get_y()],
+            Self::BlsScalar(scalar, _) => {
+                bytes[0] = 0;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::JubJubScalar(scalar, _) => {
+                bytes[0] = 1;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::AffinePoint(point, _, _) => {
+                bytes[0] = 2;
+                bytes[1..Self::serialized_size()].copy_from_slice(&point.to_bytes());
+                bytes
+            }
+        }
+    }
+
+    /// Generate a [`PublicInput`] structure from it's byte representation.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, CircuitErrors> {
+        if bytes.len() != Self::serialized_size() {
+            return Err(CircuitErrors::InvalidPublicInputBytes.into());
+        } else {
+            let mut array_bytes = [0u8; 32];
+            array_bytes.copy_from_slice(&bytes[1..Self::serialized_size()]);
+            match bytes[0] {
+                0 => Ok(Self::BlsScalar(
+                    Option::from(BlsScalar::from_bytes(&array_bytes))
+                        .ok_or_else(|| CircuitErrors::InvalidPublicInputBytes.into())?,
+                    0,
+                )),
+                1 => Ok(Self::JubJubScalar(
+                    Option::from(JubJubScalar::from_bytes(&array_bytes))
+                        .ok_or_else(|| CircuitErrors::InvalidPublicInputBytes.into())?,
+                    0,
+                )),
+                2 => Ok(Self::AffinePoint(
+                    Option::from(JubJubAffine::from_bytes(array_bytes))
+                        .ok_or_else(|| CircuitErrors::InvalidPublicInputBytes.into())?,
+                    0,
+                    0,
+                )),
+                _ => unreachable!(),
+            }
         }
     }
 
     /// Returns the positions that of a PublicInput struct
-    pub fn pos(&self) -> Vec<usize> {
+    fn pos(&self) -> [usize; 2] {
```

Comment:
Is this needed for the serialization? Or, in other world, why this can't be a tuple? I don't mind but for points (x, y) seems more logical to me have a tuple, but it's fine either way

Reply:
This is not needed for serialization. Is only used inside of some Trait fns. So for now I would keep it as is.

Reply:
That's the point, if it's not needed for serialization I don't see the point to have an array of two instead of a tuple for points. But as said, I don't mind. Let's keep as is for the time being.